advisor/src/main/kotlin/advisors/VulnerableCode.kt

```diff
@@ -0,0 +1,168 @@
+    class Factory : AbstractAdvisorFactory<VulnerableCode>("VulnerableCode") {
+        override fun create(config: AdvisorConfiguration) = VulnerableCode(advisorName, config)
+    }
+
+    companion object {
+        /** The number of elements to request at once in a bulk request. */
+        private const val BULK_FETCH_SIZE = 100
+
+        /** A severity value used if no information is provided by VulnerableCode. */
+        private const val SEVERITY_UNDEFINED = -1f
+
+        /** A regular expression pattern to extract the ID of a vulnerability from its URI. */
+        private val vulnerabilityIdPattern = Regex(".*/api/vulnerabilities/([^/]+)/")
+
+        /**
+         * Try to extract the (internal VulnerableCode) ID from the given [url] of a vulnerability. Throw an
+         * exception if the URL has an unexpected form.
+         */
+        private fun extractVulnerabilityId(url: String): String =
+            vulnerabilityIdPattern.matchEntire(url)?.groups?.get(1)?.value
+                ?: throw IOException("Unexpected URL of a vulnerability: $url.")
+    }
+
+    private val vulnerableCodeConfiguration = config as VulnerableCodeConfiguration
+
+    override suspend fun retrievePackageVulnerabilities(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val service = VulnerableCodeService.create(vulnerableCodeConfiguration.vulnerableCodeServerUrl)
+
+        val startTime = Instant.now()
+
+        val components = packages.map { it.purl }
+
+        @Suppress("TooGenericExceptionCaught")
+        return try {
+            val packagesToVulnerabilityRefs = mutableMapOf<String, VulnerableCodeService.Vulnerabilities>()
+
+            components.chunked(BULK_FETCH_SIZE).forEach { component ->
+                val requestResults =
+                    service.getPackageVulnerabilities(VulnerableCodeService.PackagesWrapper(component))
+                packagesToVulnerabilityRefs += requestResults
+            }
+
+            val packagesToVulnerabilities = fetchVulnerabilityDetails(packagesToVulnerabilityRefs, service)
+
+            val endTime = Instant.now()
+
+            packages.mapNotNull { pkg ->
+                packagesToVulnerabilities[pkg.id.toPurl()]?.let { vulnerabilities ->
+                    pkg to listOf(
+                        AdvisorResult(
+                            vulnerabilities,
+                            AdvisorDetails(advisorName),
+                            AdvisorSummary(startTime, endTime)
+                        )
+                    )
+                }
+            }.toMap()
+        } catch (e: Exception) {
+            e.showStackTrace()
+
+            val now = Instant.now()
+            val failedResults = listOf(
+                AdvisorResult(
+                    vulnerabilities = emptyList(),
+                    advisor = AdvisorDetails(advisorName),
+                    summary = AdvisorSummary(
+                        startTime = now,
```

Comment:
You can use the real `startTime` here.
Or is there a reason to not do this in the error case?